File: java/jeromq/app/src/main/java/com/example/jeromq/ZeroMQServer.java
package com.example.jeromq;

import android.os.Handler;

import org.zeromq.ZMQ;

public class ZeroMQServer implements Runnable {

	public ZeroMQServer(Handler uiThread) {
		_uiThread = uiThread;
	}

	@Override
	public void run() {
		ZMQ.Context ctx = ZMQ.context(1);
		ZMQ.Socket rep = ctx.socket(ZMQ.REP);
		rep.bind("tcp://127.0.0.1:5555");

		while (!Thread.currentThread().isInterrupted()) {
			byte[] msg = rep.recv(0);
			_uiThread.sendMessage(
				Util.bundledMessage(_uiThread, new String(msg)));
			rep.send(new String(Util.reverseInPlace(msg)), 0);
		}

		rep.close();
		ctx.term();
	}

	private final Handler _uiThread;
}

File: kotlin/jeromq/app/src/main/java/example/kotlin/jeromq/ZeroMQServer.kt
package example.kotlin.jeromq

import android.os.Handler
import org.zeromq.ZContext
import org.zeromq.ZMQ

class ZeroMQServer(private val uiThread: Handler) : Runnable {

	override fun run() {
		val ctx = ZContext(1)
		val rep = ctx.createSocket(ZMQ.REP)
		rep.bind("tcp://127.0.0.1:5555")

		while (!Thread.currentThread().isInterrupted) {
			val msg = rep.recv(0)
			uiThread.sendMessage(Util.bundledMessage(uiThread, String(msg)))
			rep.send(Util.reverse(String(msg)), 0)
		}

		rep.close()
		ctx.close()
	}
}